### PR TITLE
Add DTLS support

### PR DESCRIPTION
**Pull request by manuels**
This commit adds support for DTLS (TLS over udp connections).
Note that DTLS requires cookies to prevent DoS attacks. In this implementation this is achieved by using hmac(secret, (ip_addr, port)) using sha256 where secret is a 48 byte random string.

### CODE REVIEW

File: src/lib/libtls/tls_server.c [hunk 1]
```diff
@@ -106,14 +115,31 @@ tls_accept_socket(struct tls *ctx, struct tls **cctx, int socket)
...
 		SSL_set_app_data(conn_ctx->ssl_conn, conn_ctx);
 	}
 
-	if ((ret = SSL_accept(conn_ctx->ssl_conn)) != 1) {
+	if (tls_config_is_dtls(ctx->config)) {
+		union {
+			struct sockaddr_in s4;
+			struct sockaddr_in6 s6;
+		} client_addr = { .s6 = {} };
+		ret = DTLSv1_listen(conn_ctx->ssl_conn, &client_addr);
```
Comment (manuels) [inline src/lib/libtls/tls_server.c]: Actually, a simple `SSL_accept()` should work here, too. But calling `DTLSv1_listen()` has some additional impact on packet counting.

**Comment by busterb** [comment]
Thanks for the patch. Do you have a program in mind that will use this support? I don't think we want to add features at-will without there being a clear reason/test case for them, and so far there are not a lot of programs in OpenBSD base that use DTLS in the first place.

Also, I spotted some style issues, please check http://www.openbsd.org/cgi-bin/man.cgi/OpenBSD-current/man9/style.9?query=style&arch=i386

You should bring this up for discussion at tech@openbsd.org, since it is not just a -portable specific change.

**Comment by manuels** [comment]
Hi, well there is lots of VPN software using DTLS (e.g. [[1]](http://campagnol.sourceforge.net/)).
I'm sorry for the style issues, I will have a look at that.

I already [sent a mail to tech@openbsd.org](https://marc.info/?l=openbsd-tech&m=142115520322259&w=2), however, there were no reactions on the mailing list :(

**Comment by busterb** [comment]
I checked around a little and there is not a lot of interest in DTLS yet. DTLS as used in the OpenConnect VPN uses a Cisco-specific non-standard variant. It also seems to have a high rate of being somewhat buggy. Let's leave this open and revisit during the next release cycle. @4a6f656c @tedu  does this sound reasonable?

**Comment by 4a6f656c** [comment]
I'm reluctant to add DTLS without an existing use case/tool - the existing implementations in libtls are well used by ftp(1) and httpd(8) in base (now along with ntpd(8), spamd(8) and syslogd(8)) and that drove the design. Without an application that uses libtls DTLS (even one in ports would be potentially fine), we're simply adding an unused and untested API.

**Comment by bob-beck** [comment]
Manuels - do you have a portential use case for this? Again - we don't want to support this without real users of it, as Joel as suggested

**Comment by busterb** [comment]
This patch some work to get it working again. I'm going to close this PR, but might repost at tech@ if I get it running with nc(1).

**Pull request closed without merging**